### DISCUSSION TITLE
Best way to exclude admin section from client js files?

### DISCUSSION THREAD

**Discussion by nullday0** [remix-run/remix#7912] [Q&A] [answered]
The goal is to extract admin section components and everything from main js bundle.

I would like to extract admin pieces into separate js files with some prefix (say, `admin-<hash>.chunk.js`), and then limit access to js files with such prefix on backend side (using express or whatever) using sessions.

The question is what tools can I use to achieve this?
Seems like there is code splitting on route level, but how to manage each route's build filename, to add that admin prefix?

This is a security issue to me, for example, there could be some admin pages having some sensitive internal directions/instructions. And as it's static data, it will be packed into final js bundle, and some bad guys could reverse engineer public js, get admin paths and fetch these chunks of JS, and basically read all these templates with sensitive info.
I can always deploy a separate admin app, but it's not an option for some projects.

Any input will be appreciated, thanks.

**Comment by sergiodxa** [comment, marked as answer]
After you build your app, you will have inside your `public/build` folder a file named `manifest-<hash>.js`, this file contains the information about the routes of your app, what's the entry module of that route and what other files it imports.

The file looks something like this
```js
window.__remixManifest = {
  entry: { module: "path", imports: ["path1", "path2"] },
  routes: {
    root: { ... },
    admin: {
      id: "admin",
      parentId: "root", // empty string for root, every other route will have at least root as value
      path: "admin",
      module: "path",
      imports: ["path1", "path2", ...],
      hasAction: false,
      hasLoader: true,
      hasErrorBoundary: true,
    }
    ...
  },
  version: "hash",
  url: "url that points to this file",
}
```

So you could, in your HTTP server, read this file, find all your admin related route, find the path (which is the entry JS file for each route) and every import (note that some of these imports could be shared with non-admin routes) and the add auth protection to the files.

This way you don't need to rename the files or anything, just find what files needs to be protected based on the routes they belong to, and protect them.

---

I also recommend you to try send any static data from the loader, since loaders a server-side only you can use them to pass private information from the server to the client, this way you may not need to protect the JS files at all since they will not contain any sensitive information anymore.

**Reply by nullday0** [reply]
Thanks, this seems like what I was looking for.
P.S. Also thanks for all your articles about remix in general.
